test/Npgsql.Tests/FunctionTests.cs

```diff
@@ -177,6 +177,28 @@ public async Task CommandBehavior_SchemaOnly_support_function_call()
         Assert.AreEqual(0, i);
     }
 
+    [Test, IssueLink("https://github.com/npgsql/npgsql/issues/5820")]
+    public async Task Output_param_cast_error()
+    {
+        await using var conn = await OpenConnectionAsync();
+        var function = await GetTempFunctionName(conn);
+        await conn.ExecuteNonQueryAsync(@$"
+CREATE FUNCTION {function} (INOUT param_in int4, OUT param_out interval) AS $$
+BEGIN
+    param_out = interval '5 years';
+END
+$$ LANGUAGE plpgsql");
+        await using var cmd = new NpgsqlCommand(function, conn);
+        cmd.CommandType = CommandType.StoredProcedure;
+        cmd.Parameters.Add(new NpgsqlParameter("param_in", DbType.Int32)
+        {
+            Direction = ParameterDirection.InputOutput,
+            Value = 1
+        });
+        Assert.ThrowsAsync<InvalidCastException>(cmd.ExecuteNonQueryAsync);
```

Comment:
Just for my mental picture, what was the reason we would throw here originally? Don't we skip over parameters/outputs that we haven't defined in the collection?

Reply:
There is a weird behavior we have in that we try to populate unmatched parameters with unmatched values. We have #2252 open to investigate and potentially remove this.

Reply:
I modified the test to also have out param in case we someday remove that behavior

Reply:
Yes I wrote about the behavior in https://github.com/npgsql/npgsql/pull/5645/files#diff-9508a102a198d47c59b1dc9bd1da180a57e6a47d800e6b9639609a27539e4980R305-R311